File: javascore/bmv/bmv/src/main/java/foundation/icon/btp/bmv/edgeware/IBMV.java
package foundation.icon.btp.bmv.edgeware;

import foundation.icon.btp.lib.BMVStatus;
import foundation.icon.btp.lib.mta.MTAStatus;

import score.Address;

public interface IBMV {
    /**
     * get encode of merkle tree accumulator
     */
    String mta();

    /**
     * get address of bmc
     */
    Address bmc();

    /**
     * net address that bmv verify
     */
    String netAddress();

    /**
     * list addresses of validator
     */
    String validators();

    /**
     * get status of BMV
     */
    BMVStatus getStatus();

    /**
     * handle verify message from BMC and return list of event message
     */
    String[] handleRelayMessage(String bmc, String prev, long seq, String msg);
}

File: rust/near/bsh-v2/bsh-periphery/src/interfaces/ibmv.rs
use near_sdk::ext_contract;

#[ext_contract(ibmv)]
pub trait Ibmv {
    /**
       @return        Base64 encode of Merkle Tree
    */
    fn get_mta(&self) -> String;

    /**
       @return        Connected BMC address
    */
    fn get_connected_bmc(&self) -> AccountId;

    /**
       @return        Network address of the blockchain
    */
    fn get_net_address(&self) -> AccountId;

    /**
       @return        Hash of RLP encode from given list of validators
       @return        List of validators' addresses
    */
    fn get_validators(&self) -> (CryptoHash, Vec<AccountId>);

    /**
       @notice Used by the relay to resolve next BTP Message to send.
               Called by BMC.
       @return        Height of MerkleTreeAccumulator
       @return        Offset of MerkleTreeAccumulator
       @return        Block height of last relayed BTP Message
    */
    fn get_status(&self) -> (u128, usize, u128);

    /**
       @notice Decodes Relay Messages and process BTP Messages.
               If there is an error, then it sends a BTP Message containing the Error Message.
               BTP Messages with old sequence numbers are ignored. A BTP Message contains future sequence number will fail.
       @param bmc     BTP Address of the BMC handling the message
       @param prev    BTP Address of the previous BMC
       @param seq     Next sequence number to get a message
       @param msg     Serialized bytes of Relay Message
       @return        List of serialized bytes of a BTP Message
    */
    fn handle_relay_message(
        &mut self,
        bmc: AccountId,
        prev: AccountId,
        seq: u128,
        msg: String,
    ) -> Vec<Vec<u8>>;
}
